qpmodel/PhysicalNode.cs

```diff
@@ -1113,15 +1143,36 @@ public override string Exec(Func<Row, string> callback)
             });
 
             // special handling for emtpy resultset and last row
-            if (!context.option_.optimize_.use_codegen_)
+            if (context.option_.optimize_.use_codegen_)
+            {
+                if (logic.groupby_ is null)
+                    s += $@"
+                    if (curGroupRow{_} is null)
+                    {{
+                        Row r{_} = {_physic_}.HandleEmptyResult(context);
+                        if (r{_} != null)
+                        {{
+                            {callback(null)}
+                        }}
+                    }}
+                    ";
+                s += $@"
+                if (curGroupKey{_} != null && !context.stop_)
+                {{
+                    var keys{_} = curGroupKey{_};
+                    Row aggvals{_} = curGroupRow{_};
+                    {FinalizeAGroupRow(context, null, null, callback)}
+                }}";
+            }
+            else
             {
                 if (logic.groupby_ is null && curGroupRow is null)
                 {
                     Row row = HandleEmptyResult(context);
                     if (row != null)
                         callback(row);
                 }
-                if (curGroupKey != null)
+                if (curGroupKey != null && !context.stop_)
```

Comment:
I am having some concern about this change. It does not seem as clean for stop to appear in aggregate node.

Reply:
We may want to stop current execution for some reasons: 1) LIMIT gets enough data; 2) query cancel due to user or abort due to internal error. context.stop_ is designed for the 1st one. The second one will be handled by try/catch exception.

Usually we check stop_ flag on top of each loop thus we won't waste too much before stop.